modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -70,48 +168,129 @@ object GlobalL0Service {
-                .map { lastOrdinal =>
-                  val nextOrdinal = lastStoredOrdinal.next
-                  val lastOrdinalCap = lastOrdinal.value.value
-                    .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
-
-                  nextOrdinal.value.value to lastOrdinalCap
-                }
-                .map(_.toList.map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o))))
-                .flatMap { ordinals =>
-                  (ordinals, List.empty[Hashed[GlobalIncrementalSnapshot]]).tailRecM {
-                    case (ordinal :: nextOrdinals, snapshots) =>
-                      l0GlobalSnapshotClient
-                        .get(ordinal)(l0Peer)
-                        .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
-                        .map(s => (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
-                        .handleErrorWith { e =>
-                          logger
-                            .warn(e)(s"Failure pulling snapshot with ordinal=$ordinal")
-                            .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
-                        }
-
-                    case (Nil, snapshots) =>
-                      Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
-                  }
-                }
-            }
-
-            pulled.map(_.asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)])
+            for {
+              l0Peer <- globalL0ClusterStorage.getRandomPeer
+              lastOrdinal <- l0GlobalSnapshotClient.getLatestOrdinal.run(l0Peer)
+              pulled <- pullSnapshots(l0Peer, lastStoredOrdinal.next, lastOrdinal)
+            } yield pulled.asRight[LatestSnapshotTuple]
           }
         }.handleErrorWith { e =>
-          logger.warn(e)(s"Failure pulling global snapshots!") >>
-            Applicative[F].pure(
-              List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
+          logger
+            .warn(e)("Failure pulling global snapshots")
+            .as(
+              List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[LatestSnapshotTuple]
             )
         }
+
+      private def pullSnapshots(
+        l0Peer: L0Peer,
+        nextOrdinal: SnapshotOrdinal,
+        lastOrdinal: SnapshotOrdinal
+      ): F[List[Hashed[GlobalIncrementalSnapshot]]] = {
+        val lastOrdinalCap = lastOrdinal.value.value
+          .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
+
+        val ordinals = (nextOrdinal.value.value to lastOrdinalCap).toList
+          .map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o)))
+
+        type B = Hashed[GlobalIncrementalSnapshot]
+        type S = (List[SnapshotOrdinal], Chain[B])
+        type R = List[B]
```

Comment:
can you use `Success`, `Result` and `Agg` for `tailRecM` types? this is what we already have in the codebase

![image](https://github.com/Constellation-Labs/tessellation/assets/4959931/04817bac-2b69-4606-8d46-8ae54b5c1aae)
![image](https://github.com/Constellation-Labs/tessellation/assets/4959931/5b004091-c0e1-458e-836e-0ad69ac27b33)
![image](https://github.com/Constellation-Labs/tessellation/assets/4959931/6984c605-e1cb-4652-9aca-09d2e841e32e)
![image](https://github.com/Constellation-Labs/tessellation/assets/4959931/81afee89-ad1d-418f-a481-c68a90ce0843)

Reply:
done